src/serialize.rs

```diff
@@ -79,40 +88,148 @@ fn decode_size(f: &mut dyn Read, initial_b: u8) -> std::io::Result<usize> {
     }
     // need to convert size_blob to an int
     let mut v: usize = 0;
+    if size_blob.len() > 6 {
+        return Err(bad_encoding());
+    }
     for b in size_blob.iter() {
         v <<= 8;
         v += *b as usize;
     }
-    let bytes_to_read = v;
-    Ok(bytes_to_read)
+    if v >= 0x400000000 {
+        return Err(bad_encoding());
+    }
+    Ok(v)
 }
 
 pub fn node_from_stream<T: Allocator>(allocator: &T, f: &mut dyn Read) -> std::io::Result<T::Ptr> {
+    let mut values = Vec::<T::Ptr>::new();
+    let mut ops = vec![0_u8];
+
     let mut b = [0; 1];
-    f.read_exact(&mut b)?;
-    if b[0] == CONS_BOX_MARKER {
-        let v1 = node_from_stream(allocator, f)?;
-        let v2 = node_from_stream(allocator, f)?;
-        return Ok(allocator.new_pair(&v1, &v2));
-    }
-    if b[0] <= MAX_SINGLE_BYTE {
-        return Ok(allocator.new_atom(&b));
+    loop {
+        let op = ops.pop();
+        if op.is_none() {
+            break;
+        }
+        match op.unwrap() {
+            0 => {
+                f.read_exact(&mut b)?;
+                if b[0] == CONS_BOX_MARKER {
+                    ops.push(1); // cons
```

Comment:
This might be a good place to use a simple `enum` of two types with no values (which rust will abstract to something very efficient).

Reply:
I agree. I should have done that. I get the impression that a rust `Enum` is a variant though...